exporter/awskinesisexporter/internal/batch/batch.go

```diff
@@ -98,18 +108,45 @@ func (b *Batch) AddRecord(raw []byte, key string) error {
 		return ErrPartitionKeyLength
 	}
 
-	if l := len(record); l == 0 || l > b.maxRecordSize {
+	if l := len(compressed); l == 0 || l > b.maxRecordSize {
 		return ErrRecordLength
 	}
 
-	b.records = append(b.records, &kinesis.PutRecordsRequestEntry{Data: record, PartitionKey: aws.String(key)})
+	if b.useAggregation && b.aggregator.IsRecordAggregative(compressed, key) {
+		if b.aggregator.IsBatchFull(compressed, key) {
+			record, err := b.aggregator.Drain()
+			if err != nil {
+				log.Fatal(err)
+			}
+			if record != nil {
+				b.records = append(b.records, record)
+			}
+		} else {
+			b.aggregator.Put(compressed, key)
+		}
+	} else {
+		b.records = append(b.records, &kinesis.PutRecordsRequestEntry{Data: compressed, PartitionKey: aws.String(key)})
+	}
 	return nil
 }
 
+func (b *Batch) Close() {
+	if !b.useAggregation {
+		return
+	}
+	record, err := b.aggregator.Drain()
+	if err != nil {
+		log.Fatal(err)
+	}
+	b.records = append(b.records, record)
+}
+
 // Chunk breaks up the iternal queue into blocks that can be used
 // to be written to he kinesis.PutRecords endpoint
 func (b *Batch) Chunk() (chunks [][]*kinesis.PutRecordsRequestEntry) {
 	// Using local copies to avoid mutating internal data
+	b.Close()
```

Comment:
To me this is a code smell, `batch` shouldn't hold any resources open to close, and a chunk can be called multiple times without side affects.

Reply:
I removed the aggregation from the batch to the batchMarshaller